src/mbgl/style/expression/in.cpp

```diff
@@ -0,0 +1,116 @@
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/util/string.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+EvaluationResult In::evaluate(const EvaluationContext& params) const {
+    const EvaluationResult evaluatedNeedle = needle->evaluate(params);
+    const EvaluationResult evaluatedHeystack = haystack->evaluate(params);
+
+    if (!evaluatedNeedle) {
+        return evaluatedNeedle.error();
+    }
+    if (!evaluatedHeystack) {
+        return evaluatedHeystack.error();
+    }
+
+    type::Type evaluatedNeedleType = typeOf(*evaluatedNeedle);
+    if (!isComparableRuntimeValue(evaluatedNeedleType)) {
+        return EvaluationError{"Expected first argument to be of type boolean, string or number, but found " +
+                               toString(evaluatedNeedleType) + " instead."};
+    }
+
+    type::Type evaluatedHeystackType = typeOf(*evaluatedHeystack);
+    if (!isSearchableRuntimeValue(evaluatedHeystackType)) {
+        return EvaluationError{"Expected second argument to be of type array or string, but found " +
+                               toString(evaluatedHeystackType) + " instead."};
+    }
+
+    if (evaluatedNeedleType == type::Null || evaluatedHeystackType == type::Null) {
+        return EvaluationResult(false);
+    }
+
+    if (evaluatedHeystackType == type::String) {
+        const auto heystackString = evaluatedHeystack->get<std::string>();
+        std::string needleValue = "";
+        if (evaluatedNeedleType == type::Boolean) {
+            needleValue = evaluatedNeedle->get<bool>() ? "true" : " false";
+        } else if (evaluatedNeedleType == type::String) {
+            needleValue = evaluatedNeedle->get<std::string>();
+        } else if (evaluatedNeedleType == type::Number) {
+            needleValue = std::to_string(evaluatedNeedle->get<double>());
+            needleValue.erase(needleValue.find_last_not_of('0') + 1, std::string::npos);
+            needleValue.erase(needleValue.find_last_not_of('.') + 1, std::string::npos);
+        }
+        return EvaluationResult(heystackString.find(needleValue) != std::string::npos);
+    } else {
+        const auto heystackArray = evaluatedHeystack->get<std::vector<Value>>();
+
+        bool result = false;
+        if (evaluatedNeedleType == type::Boolean) {
+            auto needleValue = evaluatedNeedle->get<bool>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        } else if (evaluatedNeedleType == type::String) {
+            auto needleValue = evaluatedNeedle->get<std::string>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        } else if (evaluatedNeedleType == type::Number) {
+            auto needleValue = evaluatedNeedle->get<double>();
+            result = find(heystackArray.begin(), heystackArray.end(), needleValue) != heystackArray.end();
+        }
```

Comment:
Value has comparison operator. Needle type comparability is checked on[ line 23 `isComparableType`](https://github.com/mapbox/mapbox-gl-native/pull/16162/files#diff-5d51b639578982530d698f5b5fd0536eR23).

Whole block can be rewritten as:

```c++
return EvaluationResult(find(heystackArray.begin(), heystackArray.end(), evaluatedNeedle) != heystackArray.end());
```

Reply:
The type of `evaluatedNeedle` is `EvaluationResult ` , it can't be compared with type `Value`.

Reply:
``` c++
    } else {
        const auto haystackArray = evaluatedHaystack->get<std::vector<Value>>();
        return EvaluationResult(std::find(haystackArray.begin(), haystackArray.end(), *evaluatedNeedle) != haystackArray.end());
    }
```